### PR TITLE
Add negative tests for constraint mapper

### PR DESCRIPTION
**Pull request by SachinAkash01**
- Add negative test cases  for Ballerina to OpenAPI constraint mapper.

Resolves: [#4945](https://github.com/ballerina-platform/ballerina-standard-library/issues/4945)
See Also: [#4788](https://github.com/ballerina-platform/ballerina-standard-library/issues/4788)

### CODE REVIEW

File: openapi-cli/src/test/java/io/ballerina/openapi/generators/openapi/NegativeConstraintTests.java [hunk 1]
```diff
@@ -49,4 +55,64 @@ public void testInvalidRegexPatterns() throws IOException {
...
+        Assert.assertTrue(errors[0].toString().contains("ERROR [invalidRegexPattern.bal:(21:20,21:20)] " +
+                "invalid char after backslash"));
+        Assert.assertTrue(errors[1].toString().contains("ERROR [invalidRegexPattern.bal:(21:22,21:22)] " +
+                "missing backslash"));
+    }
+
+    @Test(description = "When Integer constraint has float value which is invalid")
+    public void testInvalidInteger() {
+        Path ballerinaFilePath = RES_DIR.resolve("constraint-negative/invalidInteger.bal");
+        Project project = getProject(ballerinaFilePath);
```
Comment (ayeshLK) [inline openapi-cli/src/test/java/io/ballerina/openapi/generators/openapi/NegativeConstraintTests.java:110]:
```suggestion
    public void testInvalidInteger() throws ProjectException {
        Path ballerinaFilePath = RES_DIR.resolve("constraint-negative/invalidInteger.bal");
        Project project = ProjectLoader.loadProject(servicePath);
```

Check for other places as well.

File: openapi-cli/src/test/java/io/ballerina/openapi/generators/openapi/TestUtils.java [hunk 2]
```diff
@@ -84,6 +89,22 @@ public static List<OpenAPIConverterDiagnostic> compareWithGeneratedFile(OASContr
         }
     }
 
+    public static Project getProject(Path servicePath) {
```
Comment (ayeshLK) [inline openapi-cli/src/test/java/io/ballerina/openapi/generators/openapi/TestUtils.java:92]: As discussed lets remove this.

File: openapi-cli/src/test/java/io/ballerina/openapi/generators/openapi/NegativeConstraintTests.java (hunk 1 above)
Comment (SachinAkash01) [reply openapi-cli/src/test/java/io/ballerina/openapi/generators/openapi/NegativeConstraintTests.java]: Resolved

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java [hunk 3]
```diff
@@ -863,6 +863,23 @@ private boolean isConstraintAnnotation(AnnotationNode annotation) {
         return false;
     }
 
+    /*
+     * This util is used to check whether an annotation is a constraint:Date annotation.
+     * Currently, we don't have support for mapping Date constraints to OAS hence we skip them.
+     * {@link <a href="https://github.com/ballerina-platform/ballerina-standard-library/issues/5049">...</a>}
+     * Once the above improvement is completed this method should be removed!
+     */
+    private boolean isNotConstraintDate(AnnotationNode annotation) {
```
Comment (ayeshLK) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java:872]: This method does two things:
1. Check whether the current constraint is a `Date` constraint.
2. If so update the compilation context with a warning. 

This breaks SRP. Lets check only whether the current constraint is a `Date` constraint or not in this method.

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java [hunk 4]
```diff
@@ -841,6 +840,7 @@ private void extractedConstraintAnnotation(MetadataNode metadata,
         NodeList<AnnotationNode> annotations = metadata.annotations();
         annotations.stream()
                 .filter(this::isConstraintAnnotation)
+                .filter(this::isNotConstraintDate)
                 .filter(annotation -> annotation.annotValue().isPresent())
                 .forEach(annotation -> {
                     MappingConstructorExpressionNode annotationValue = annotation.annotValue().get();
```
Comment (ayeshLK) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java:846]: Relevant to the this comment [1] 
```suggestion
                .filter(annotation -> annotation.annotValue().isPresent())
                .forEach(annotation -> {
                    if (isDateConstraint(annotation)) {
                        DiagnosticMessages errorMsg = DiagnosticMessages.OAS_CONVERTOR_120;
                        IncompatibleResourceDiagnostic error = new IncompatibleResourceDiagnostic(errorMsg,
                annotation.location(), annotation.toString());
                        diagnostics.add(error);
                        return;
                    }
                    MappingConstructorExpressionNode annotationValue = annotation.annotValue().get();
```

[1] - https://github.com/ballerina-platform/openapi-tools/pull/1559/files#r1363662026

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java [hunk 5]
```diff
@@ -863,6 +863,23 @@ private boolean isConstraintAnnotation(AnnotationNode annotation) {
...
     }
 
+    /*
+     * This util is used to check whether an annotation is a constraint:Date annotation.
+     * Currently, we don't have support for mapping Date constraints to OAS hence we skip them.
+     * {@link <a href="https://github.com/ballerina-platform/ballerina-standard-library/issues/5049">...</a>}
+     * Once the above improvement is completed this method should be removed!
+     */
+    private boolean isNotConstraintDate(AnnotationNode annotation) {
+        if (!annotation.annotReference().toString().trim().equals("constraint:Date")) {
```
Comment (ayeshLK) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java:873]: Lets move `constraint:Date` into a constant.

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java (hunk 4 above)
Comment (SachinAkash01) [reply openapi-bal-service/src/main/java/io/ballerina/openapi/converter/service/OpenAPIComponentMapper.java]: Resolved!

File: openapi-bal-service/src/main/java/io/ballerina/openapi/converter/Constants.java [hunk 6]
```diff
@@ -83,6 +83,7 @@ public class Constants {
     public static final String MEDIA_TYPE = "mediaType";
     public static final String TUPLE = "tuple";
     public static final String REGEX_INTERPOLATION_PATTERN = "^(?!.*\\$\\{).+$";
+    public static final String CHECK_DATE_CONSTRAINT = "constraint:Date";
```
Comment (ayeshLK) [inline openapi-bal-service/src/main/java/io/ballerina/openapi/converter/Constants.java:86]:
```suggestion
    public static final String DATE_CONSTRAINT_ANNOTATION = "constraint:Date";
```

**Review by ayeshLK — APPROVED** [review]

**Pull request merged**